File: src/main/java/jugglinglab/jml/HandLink.java
//
// HandLink.java
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.jml;

import jugglinglab.curve.*;
import jugglinglab.util.*;

public class HandLink {
  protected int jugglernum;
  protected int handnum;
  protected JMLEvent startevent;
  protected JMLEvent endevent;
  protected VelocityRef startvelref;
  protected VelocityRef endvelref;
  protected Curve hp;

  protected boolean ismaster;
  protected HandLink[] duplicates; // if master
  protected HandLink master; // if duplicate

  public static final int NO_HAND = 0;
  public static final int LEFT_HAND = 1;
  public static final int RIGHT_HAND = 2;

  public HandLink(int jugglernum, int handnum, JMLEvent startevent, JMLEvent endevent) {
    this.jugglernum = jugglernum;
    this.handnum = handnum;
    this.startevent = startevent;
    this.endevent = endevent;
    hp = null;
    ismaster = true;
    duplicates = null;
    master = null;
  }

  public static int index(int handdescription) {
    return (handdescription == LEFT_HAND ? 0 : 1);
  }

  public int getJuggler() {
    return jugglernum;
  }

  public int getHand() {
    return handnum;
  }

  public JMLEvent getStartEvent() {
    return startevent;
  }

  public JMLEvent getEndEvent() {
    return endevent;
  }

  public VelocityRef getStartVelocityRef() {
    return startvelref;
  }

  public void setStartVelocityRef(VelocityRef vr) {
    startvelref = vr;
  }

  public VelocityRef getEndVelocityRef() {
    return endvelref;
  }

  public void setEndVelocityRef(VelocityRef vr) {
    endvelref = vr;
  }

  public void setHandCurve(Curve hp) {
    this.hp = hp;
  }

  public Curve getHandCurve() {
    return hp;
  }

  public boolean isMaster() {
    return ismaster;
  }

  @Override
  public String toString() {
    String result = null;

    Coordinate start = startevent.getGlobalCoordinate();
    result =
        "Link from (x="
            + start.x
            + ",y="
            + start.y
            + ",z="
            + start.z
            + ",t="
            + startevent.getT()
            + ") ";
    Coordinate end = endevent.getGlobalCoordinate();
    result += "to (x=" + end.x + ",y=" + end.y + ",z=" + end.z + ",t=" + endevent.getT() + ")";

    VelocityRef vr = getStartVelocityRef();
    if (vr != null) {
      Coordinate vel = vr.getVelocity();
      result += "\n      start velocity (x=" + vel.x + ",y=" + vel.y + ",z=" + vel.z + ")";
    }
    vr = getEndVelocityRef();
    if (vr != null) {
      Coordinate vel = vr.getVelocity();
      result += "\n      end velocity (x=" + vel.x + ",y=" + vel.y + ",z=" + vel.z + ")";
    }
    Curve hp = getHandCurve();
    if (hp != null) {
      Coordinate maxcoord = hp.getMax(startevent.getT(), endevent.getT());
      Coordinate mincoord = hp.getMin(startevent.getT(), endevent.getT());
      result += "\n      minimum (x=" + mincoord.x + ",y=" + mincoord.y + ",z=" + mincoord.z + ")";
      result += "\n      maximum (x=" + maxcoord.x + ",y=" + maxcoord.y + ",z=" + maxcoord.z + ")";
    } else {
      result += "\n      no handpath";
    }

    return result;
  }
}

File: src/main/kotlin/jugglinglab/jml/HandLink.kt
//
// HandLink.kt
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.jml

import jugglinglab.curve.Curve

class HandLink(var juggler: Int, var hand: Int, var startEvent: JMLEvent, var endEvent: JMLEvent) {
    var startVelocityRef: VelocityRef? = null
    var endVelocityRef: VelocityRef? = null
    var handCurve: Curve? = null

    /*
    var isMaster: Boolean = true
        protected set
     */

    private var duplicates: Array<HandLink?>? = null // if master
    private var master: HandLink? = null // if duplicate

    override fun toString(): String {
        val start = startEvent.globalCoordinate
        val end = endEvent.globalCoordinate
        var result =
            ("Link from (x=${start!!.x},y=${start.y},z=${start.z},t=${startEvent.t}) ")
        result += "to (x=${end!!.x},y=${end.y},z=${end.z},t=${endEvent.t})"

        val svr = startVelocityRef
        if (svr != null) {
            val vel = svr.velocity
            result += "\n      start velocity (x=${vel.x},y=${vel.y},z=${vel.z})"
        }
        val evr = endVelocityRef
        if (evr != null) {
            val vel = evr.velocity
            result += "\n      end velocity (x=${vel.x},y=${vel.y},z=${vel.z})"
        }
        val hp = handCurve
        if (hp != null) {
            val maxcoord = hp.getMax(startEvent.t, endEvent.t)
            val mincoord = hp.getMin(startEvent.t, endEvent.t)
            result += "\n      minimum (x=${mincoord!!.x},y=${mincoord.y},z=${mincoord.z})"
            result += "\n      maximum (x=${maxcoord!!.x},y=${maxcoord.y},z=${maxcoord.z})"
        } else {
            result += "\n      no handpath"
        }
        return result
    }

    companion object {
        @Suppress("unused")
        const val NO_HAND: Int = 0
        const val LEFT_HAND: Int = 1
        const val RIGHT_HAND: Int = 2

        @JvmStatic
        fun index(handdescription: Int) = if (handdescription == LEFT_HAND) 0 else 1
    }
}
